### ISSUE TITLE
PyTorch Support

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#843]
We've debated PyTorch versus TensorFlow a couple times already, but I think we should re-open the debate. PyTorch has been maturing and is starting to get buy-in from organizations (http://www.fast.ai/2017/09/08/introducing-pytorch-for-fastai/ for example, and I know of many good corporate teams making a transition more quietly).

I think it's time for us to add PyTorch support into DeepChem head. Cutting edge work on models isn't happening on TF as much these days, and I want to make sure we can keep up with the edge of the field.

To avoid causing pain to users who don't want to use PyTorch, we can use local imports to ensure that only users using PyTorch submodules actually need to have it installed.

**Comment by evanfeinberg** [comment]
+1000 to Bharath!! Let's begin discussing in this forum and offline in
Pande Lab meetings concrete steps we can take toward this goal.

**Comment by peastman** [comment]
What exactly are we discussing doing?  Most of the core classes like Dataset, Model, Metric, etc. already don't have any explicit dependence on Tensorflow, so there's no obstacle to creating new ones based on PyTorch.  But as we've discussed before, that will be very limiting for interoperability.  We've standardized on TensorGraph as the base for all models.  That makes it easy to connect up different types of modelling elements and use them together.  If we start adding new models based on PyTorch, they'll be incompatible with all TensorGraph models, which seems like a bad idea to me.

Alternatively we could convert TensorGraph (and all our other code) from Tensorflow to PyTorch.  That would be a huge project, but might be worth it in the long run if we think PyTorch is what everyone will be using in the future.

**Comment by rbharath** [comment]
My current feeling is that PyTorch will be the source of cutting edge work for probably the next couple years, while TensorFlow will dominate the enterprise and applications market for deep learning. If DeepChem is more research package than an applications project, doing a whole-scale shift might be the correct answer for our future research health even if it will be a painful transition.

**Comment by proteneer** [comment]
I don't think "cutting edge work for next couple of years" is persuasive enough of a reason given that we've *just* standardized on TensorGraph. How much of the cutting edge work is actually relevant to chemistry?

While I'm not against moving to a whole new platform per-se, I'm against a move where it's not clear what the obvious pros and cons are. Even more, the pros have to be greater than 10x the cons to justify it. Alternatively, we can try and maintain both but it's a lot tedious work. You can ask @peastman about having to support both OpenCL/CUDA/CPU-platform for OpenMM :)

**Comment by peastman** [comment]
The important questions are, what users are we trying to serve by making this change, and what problems are we trying to solve for them?

Most "DeepChem programs" either don't directly use Tensorflow at all, or they only reference it incidentally, such as to specify an activation function.  If we fully convert DeepChem over to PyTorch, it will make very little difference to the people who write those program.  The code they write will be mostly unchanged.

So the main effect of this decision is on DeepChem developers, not users.  Will the initial cost of making the conversion be outweighed in the long term by easier development?  Will this make it easier or harder to attract new developers?  Do the two packages have equivalent feature sets, or will one of them give us substantially more tools to work with?

Of course, that assumes they both work equally well.  If one of them is faster or less buggy or uses less memory, that will affect all users even if they never see the code directly.

**Comment by rbharath** [comment]
I think this tweet makes the case for PyTorch well. Check out the meme within https://twitter.com/karpathy/status/909592435440869376

The issue is that there are models easy to construct in PyTorch that are nearly impossible in TensorFlow. For example, while layers will continue to make sense, TensorGraph in it's present form may not since we won't need to only define one graph (could define a unique graph at each time step in PyTorch). Plus, by all reports, debugging models is dramatically easier in PyTorch than TensorFlow since all standard python tools work (no more weird TF stack traces). I was on the fence for a long while, but I'm becoming convinced that PyTorch style is the winner in the long run.

**Comment by Dgelemi** [comment]
I think it is possible to find people advocating for both on twitter, see this thread from F Chollet, https://twitter.com/fchollet/status/906616572898304001

**Comment by evanfeinberg** [comment]
There are definitely strong arguments to both sides. Since this is a complex and subjective matter, I recommend we schedule a conference call to those contributors that have been / will be heavily invested in such a proposed transition. How is this Friday?

**Comment by rbharath** [comment]
Let's discuss on here in more depth. Large conference calls aren't great for complicated technical issues with speakers from different orgs involved.

There's also no rush here. I'm a big fan of decision by PR; if someone can make plausible ports of DeepChem layers into PyTorch in a PR, that really helps make the decision.

**Comment by rbharath** [comment]
The votes by PR seem to be coming in fast and furious for TensorFlow judging by all the bugfixes and improvements :-). I think we're probably going in on TensorFlow for at least the next few releases.

Would people be opposed to adding a little experimental folder in `deepchem/models` for pytorch models? I'm mainly thinking we can move in some contrib pytorch models (with local imports to avoid user issues) and hook them up to tests. It would be a good to get a first hand sense of pytorch stability in case we ever want to re-open the discussion in future.

**Comment by peastman** [comment]
If it's experimental, I'd suggest keeping it in contrib.

**Comment by rbharath** [comment]
Is there a good way we could hook contrib code up to Travis? Much harder to keep track of code stability without the automated test suite.

**Comment by rbharath** [comment]
The discussion in https://github.com/deepchem/deepchem/issues/535 makes me think we should seriously have some PyTorch experimental code in DeepChem itself. The models we've been implementing have been getting increasingly sophisticated and we're running into the limits of our abstractions. From reports, PyTorch excels at complex models and I think we should evaluate its suitability as a DeepChem backend seriously to see if the hype holds up.

Note I'm not saying we should transition away from TensorGraph. I think we should stick to TensorGraph for the upcoming 1.3 and maybe even 1.4 releases. But I think we should start experimenting with PyTorch now to decide whether it might be something for 1.5 or later.

**Comment by patrickhop** [comment]
From my experience, working in PyTorch is much more productive, especially for models that do not naturally amend themselves to a static graph. I'm probably going 100% PyTorch moving forward.

**Comment by rbharath** [comment]
Going to close this issue. I think we're standardized on TensorFlow for the foreseeable future, especially given work on the new TF-eager mode.
